### PR TITLE
Added Compatibility Matrix to Upgrade Guide

### PR DESCRIPTION
**Pull request by SwastikTripathi**
As mentioned in #6044 I'm Compatibility Matrix should be added in Upgrade Guide.

### CODE REVIEW

**Comment by SwastikTripathi** [comment]
Hello again @kmaehashi 
As you said, list tables are much more maintainable. 
I've deleted the earlier changes, and made a new table.

And as @toslunar advised, the footnotes are separate.

Please review

File: docs/source/upgrade.rst [hunk 1]
```diff
@@ -360,3 +360,110 @@ Changed Behavior of count_nonzero Function
...
+     - 7.0~8.1
+     - 3.5~3.9
+     - 1.16~1.20
+     - 1.3~1.6
+     - NumPy 1.19 & SciPy 1.5
+     - v8.6.0
+   * - v7
+     - 3.0~8.x
+     - 8.0~11.0
+     - 2.x (*2)
```
Comment (kmaehashi) [inline docs/source/upgrade.rst:422]:
```suggestion
     - 2.x [2]_
```

File: docs/source/upgrade.rst [hunk 2]
```diff
@@ -360,3 +360,110 @@ Changed Behavior of count_nonzero Function
...
+     - 1.3~2.2
+     - 4.0~7.1
+     - 2.7 & 3.4~3.6
+     - 1.9~1.14
+     - (not specified)
+     - (not specified)
+     - v4.5.0
+
+.. [*] (*1) CUDA Compute Capability
+.. [*] (*2) Hightly experimental support with limited features
```
Comment (kmaehashi) [inline docs/source/upgrade.rst:469]:
```suggestion
.. [1] CUDA Compute Capability
.. [2] Highly experimental support with limited features.
```

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: docs/source/upgrade.rst [hunk 3]
```diff
@@ -360,3 +360,110 @@ Changed Behavior of count_nonzero Function
...
+     - 10.2~
+     - 4.0~
+     - 1.3~
+     - 2.8~
+     - 7.6~
+     - 3.7~
+     - 1.18~
+     - 1.4~
+     - NumPy 1.21 & SciPy 1.7
+     - latest
```
[the remark this answers is not in the document]
Comment (SwastikTripathi) [reply docs/source/upgrade.rst:394]: It's my fault I didn't notice the links😁
Thanks. Will do

**Comment by kmaehashi** [comment]
We are during code freeze, but if we don't merge this PR before the branch cut we will need to update the matrix table again, so let me merge this one now.

**Comment by leofang** [comment]
Nice work @SwastikTripathi!

**Comment by SwastikTripathi** [comment]
Got to learn so much. This community is awesome. And special thanks to @kmaehashi for constant support😊

**Pull request merged**